### PR TITLE
Implementing rANS4x8 decoder toward CRAM support

### PR DESCRIPTION
**Pull request by athos**
This PR consists of two main changes:

- Implementation of the rANS4x8 decoder
- ITF-8/LTF-8 decoder, which is required for the rANS4x8 decoder implementation

The [CRAM codec specification](https://github.com/samtools/hts-specs/blob/346a94a9980b0105e926a019b4e62fa1b9e30910/CRAMcodecs.pdf) illustrates the rANS4x8 algorithm (it also describes ITF-8/LTF-8 decoding). This PR is basically a naive implementation of it, and almost no performance tuning has been done yet, which I think will be future work.

Also, I have prepared some test resources for the rANS4x8 decoder:
- `uncompressed_order[01].dat`: Basically random byte sequences, but made somewhat realistic data
  - `uncompressed_order0.dat`: Random sequence of bytes between 0 to 2
  - `uncompressed_order1.dat`: Random sequence of ITF-8 integers representing BAM flags
- `compressed_order[01].dat.rans4x8`: Encoded from the corresponding `uncompressed_order[01].dat` by an existing rANS encoder

### CODE REVIEW

**Comment by athos** [comment]
I've tweaked some test code and resources for better test coverage.

File: test/cljam/io/cram/codecs/rans4x8_test.clj [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            [clojure.test :refer [deftest is testing]])
+  (:import [java.io FileInputStream]
+           [java.nio ByteBuffer]
+           [java.util Arrays]))
+
+(deftest read-frequencies0-test
+  ;; cited from https://github.com/samtools/hts-specs/blob/346a94a9980b0105e926a019b4e62fa1b9e30910/CRAMcodecs.tex#L389-L394
+  (let [arr (byte-array [0x61      0x87 0x47
+                         0x62 0x02 0x82 0xe8
+                         #_>       0x81 0x74
```
Comment (athos) [inline test/cljam/io/cram/codecs/rans4x8_test.clj:14]: These `#_>` are just for preventing Clojure formatters from "fixing" indentation for these lines.

**Review by alumi — APPROVED** [review]
Thank you for working on CRAM support, @athos!
I have reviewed the implementation and tests you added this time by reading the specifications for itf8 and rANS. I believe the implementation has been done correctly. 👍
Implementing codecs can be challenging, so thank you for your hard work. Please continue with the remaining CRAM specification implementation! Thank you!

File: src/cljam/io/cram/codecs/rans4x8.clj [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      state)))
+
+(defn- decode0 [bb ^long n-out]
+  (let [freqs (read-frequencies0 bb)
+        cum-freqs (cumulative-frequencies freqs)
+        states (doto (int-array 4)
+                 (aset 0 (int (bb/read-uint bb)))
+                 (aset 1 (int (bb/read-uint bb)))
+                 (aset 2 (int (bb/read-uint bb)))
+                 (aset 3 (int (bb/read-uint bb))))
```
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/cram/codecs/rans4x8.clj]: Thanks for the suggestion! Done at bc6d442 👍

**Comment by ToshimitsuArai** [comment]
LGTM, Thanks for working 👍

**Review by ToshimitsuArai — APPROVED** [review]

**Pull request merged**